### ISSUE TITLE
Need a way to capture stack traces in transpiled Dart code

### ISSUE THREAD

**Issue by mhevery** [angular/angular#1407]
```javascript
try {
  ...
} catch (e) {
  console.log(e, e.stack);
   throw e;
}
```

should transpile to

```dart
try {
  ...
} catch (e, e_stack) {
  console.log(e, e_stack);
  rethrow;
}
```

- [x] detect `e.stack` replace with `e_stack` + add `e_stack` to `catch`
- [x] detect rethrow `throw e` replace with `rethrow`;

**Comment by matanlurey** [comment]
How do you propose fixing places like so:

```dart
    try {
      d0 = length > 0 ? this._getByDependency(deps[0], binding.key) : null;
      d1 = length > 1 ? this._getByDependency(deps[1], binding.key) : null;
      d2 = length > 2 ? this._getByDependency(deps[2], binding.key) : null;
      d3 = length > 3 ? this._getByDependency(deps[3], binding.key) : null;
      d4 = length > 4 ? this._getByDependency(deps[4], binding.key) : null;
      d5 = length > 5 ? this._getByDependency(deps[5], binding.key) : null;
      d6 = length > 6 ? this._getByDependency(deps[6], binding.key) : null;
      d7 = length > 7 ? this._getByDependency(deps[7], binding.key) : null;
      d8 = length > 8 ? this._getByDependency(deps[8], binding.key) : null;
      d9 = length > 9 ? this._getByDependency(deps[9], binding.key) : null;
    } catch (e) {
      if (e is ProviderError) e.addKey(binding.key);
      throw e;
    }
```

AFAIK, there is no way to "re-throw" the stack.

**Comment by mhevery** [comment]
@matanlurey updated the bug.

**Comment by matanlurey** [comment]
:heart:

**Comment by yjbanov** [comment]
A couple of corrections:

- I think the title of the issue should be "Need a way to capture stack traces in transpiled Dart code".
- The field on JS Error object is `stack`, not `stacktrace`.

# Background

## Dart

There are two ways to capture stack traces in Dart:

### Syntactic

You can throw and catch any object, but then not every object has a stack getter. Dart has special syntax for that:

```dart
try {
  throw 'Can strings have stacks?';
} catch(e, stack) {
  print('Why yes, they can: $stack');
}
```

### Error

The [Error](https://api.dartlang.org/apidocs/channels/stable/dartdoc-viewer/dart:core.Error) class has a `stackTrace` getter:

```dart
try {
  throw new Error();
} catch(e) {
  print(e.stackTrace);
}
```

## JavaScript

If you throw anything but Error, you lose the stack trace. Error has field `stack`:

```js
try {
  throw 'Can strings have stacks?';
} catch(e, stack) {
  print('Nope.');
}

try {
  throw new Error();
} catch(e) {
  print(e.stack);
}
```

# Options

There are two more options we could consider (in addition to the proposal above):

## Only allow throwing Error

We can enforce that all thrown exceptions in Angular code extend Error. This way we can keep the classic syntax and not lose stack traces. I believe most of our exceptions stem from `BaseException` already. Many, if not all, `dart:core` errors already extend Error, e.g. RangeError, ArgumentError, StateError. So we should be covered there. We just need a facade like this:

Dart:

```dart
class ErrorWrapper {
  static StackTrace stacktrace(Error err) => err.stackTrace;
}
```

JS:

```js
export class ErrorWrapper {
  static stacktrace(err:Error) { // dunno what stack's type is
    return err.stack;
  }
}
```

## Wrap all `throw` expressions in Error

I think this is what Dart does behind the scenes. We could do the same in all Angular 2 TypeScript code. The trick is to transpile this:

```js
throw 'Can strings have stacks?';
```

To this:

```js
throw new Error('Can strings have stacks?');
```

**Comment by tbosch** [comment]
Yegor, didn't know this!
So the semantics of Dart and Js are pretty similar, except for the
fieldname on the Error class.
I would vote for just the ErrorWrapper, but not changing the JS semantics
during transpile, as we don't do this anywhere else yet...

**Comment by mprobst** [comment]
Not currently working on this, but we can pick it up later.

**Comment by yjbanov** [comment]
@tbosch If one pretends that `catch(error, stack)` syntax does not exist, then yes, semantics are the same.

**Comment by alexeagle** [comment]
I just released ts2dart 0.5.3 which includes the fix in angular/ts2dart#153

**Comment by yjbanov** [comment]
Thanks, Alex! Let's close this. https://github.com/angular/angular/issues/1962 should do the rest.
